// @node-projects/web-component-designer/dist/elements/services/undoService/UndoService.js
import { ChangeGroup } from "./ChangeGroup.js";
/*
 * Manages a stack of available undo/redo actions
 */

export class UndoService {
  constructor() {
    this._undoStack = [];
    this._redoStack = [];
    this._transactionStack = [];
  }

  openGroup(title, affectedItems) {
    let t = new ChangeGroup(title, affectedItems, t => this.commitTransactionItem(t), t => this.abortTransactionItem(t));

    this._transactionStack.push(t);

    return t;
  }

  commitTransactionItem(transactionItem) {
    let itm = this._transactionStack.pop();

    if (itm !== transactionItem) {
      this.clear();
      throw "UndoService - Commited Transation was not the last";
    }

    if (itm.undoStack.length) this._undoStack.push(itm);
  }

  abortTransactionItem(transactionItem) {
    let itm = this._transactionStack.pop();

    if (itm !== transactionItem) {
      this.clear();
      throw "UndoService - Aborted Transation was not the last";
    }

    itm.undo();
  }

  execute(item) {
    if (this._transactionStack.length == 0) {
      item.do();

      this._undoStack.push(item);

      this._redoStack = [];
    } else {
      this._transactionStack[this._transactionStack.length - 1].execute(item);
    }
  }

  clear() {
    this._undoStack = [];
    this._redoStack = [];
    this._transactionStack = [];
  }

  undo() {
    if (!this.canUndo()) return;
    if (this._transactionStack.length != 0) throw "Cannot Undo while transaction is running";

    let item = this._undoStack.pop();

    try {
      item.undo();

      this._redoStack.push(item);
    } catch (err) {
      this.clear();
      throw err;
    }
  }

  redo() {
    if (!this.canRedo()) return;
    if (this._transactionStack.length != 0) throw "Cannot Redo while transaction is running";

    let item = this._redoStack.pop();

    try {
      item.do();

      this._undoStack.push(item);
    } catch (err) {
      this.clear();
      throw err;
    }
  }

  canUndo() {
    return this._undoStack.length > 0;
  }

  canRedo() {
    return this._redoStack.length > 0;
  }

}
