// @node-projects/web-component-designer/dist/elements/services/stylesheetService/CssTreeStylesheetService.js
import { calculate as calculateSpecifity } from "./SpecificityCalculator.js";
import { AbstractStylesheetService } from "./AbstractStylesheetService.js";
export class CssTreeStylesheetService extends AbstractStylesheetService {
    _stylesheets = new Map();
    setStylesheets(stylesheets) {
        if (stylesheets != null) {
            this._stylesheets = new Map();
            for (let stylesheet of stylesheets) {
                try {
                    this._stylesheets.set(stylesheet.name, {
                        stylesheet: stylesheet,
                        ast: window.csstree.toPlainObject((window.csstree.parse(stylesheet.content, { positions: true, parseValue: false })))
                    });
                }
                catch (err) {
                    console.warn("error parsing stylesheet", stylesheet, err);
                }
            }
            this.stylesheetsChanged.emit();
        }
        else {
            this._stylesheets = null;
        }
    }
    getStylesheets() {
        let stylesheets = [];
        for (let item of this._stylesheets) {
            stylesheets.push(item[1].stylesheet);
        }
        ;
        return stylesheets;
    }
    /* Section covers the retrieval of rules and declarations */
    getAppliedRulesInternal(designItem) {
        let styles = [];
        for (let item of this._stylesheets) {
            if (!item[1].ast || !this.astHasChildren(item[1].ast))
                continue;
            styles = styles.concat(Array.from(this.rulesFromAST(item[1].ast, item[1].stylesheet.content, item[0], designItem)));
        }
        ;
        return styles;
    }
    getAppliedRules(designItem) {
        let rules = this.getAppliedRulesInternal(designItem);
        if (!rules || rules.length == 0)
            return [];
        let retCollection = [];
        for (let rule of rules) {
            retCollection.push({
                ...rule,
                declarations: rule.ast.block.children.map((declaration) => {
                    return {
                        name: declaration.property,
                        value: declaration.value.value,
                        important: declaration.important == true,
                        specificity: rule.specificity,
                        parent: rule,
                        ast: declaration,
                    };
                })
            });
        }
        return retCollection;
    }
    getDeclarationInternal(designItem, styleName) {
        let rules = this.getAppliedRulesInternal(designItem);
        if (!rules)
            return null;
        let declarations = [];
        for (let rule of rules) {
            let declaration = this.findDeclarationInRule(rule.ast, styleName);
            if (!declaration)
                continue;
            declarations.push({
                ast: declaration,
                parent: rule,
                name: styleName,
                value: declaration.value.value,
                important: declaration.important == true,
            });
        }
        ;
        return declarations;
    }
    getDeclarations(designItem, stlyeName) {
        let declarations = this.getDeclarationInternal(designItem, stlyeName);
        if (!declarations)
            return null;
        return this.sortDeclarations(declarations);
    }
    /* Section covers the update of rules and declarations */
    updateDeclarationValue(declaration, value, important) {
        let sourceNode = this._stylesheets.get(declaration.parent.stylesheetName);
        declaration.ast.value = window.csstree.toPlainObject(window.csstree.parse(declaration.name + ": " + value + (important ? " !important" : ""), { context: 'declaration', parseValue: false })).value;
        const old = sourceNode.stylesheet.content;
        sourceNode.stylesheet.content = window.csstree.generate(window.csstree.fromPlainObject(sourceNode.ast));
        // After generating the stylesheet, parse again (so line numbers are correct)
        sourceNode.ast = window.csstree.toPlainObject((window.csstree.parse(sourceNode.stylesheet.content, { positions: true, parseValue: false })));
        this.stylesheetChanged.emit({ name: sourceNode.stylesheet.name, newStyle: sourceNode.stylesheet.content, oldStyle: old, changeSource: 'styleupdate' });
        return true;
    }
    insertDeclarationIntoRule(rule, property, value, important) {
        let sourceNode = this._stylesheets.get(rule.stylesheetName);
        rule.ast.block.children.push(window.csstree.toPlainObject(window.csstree.parse(property + ": " + value + (important ? " !important" : ""), { context: 'declaration', parseValue: false })));
        const old = sourceNode.stylesheet.content;
        sourceNode.stylesheet.content = window.csstree.generate(window.csstree.fromPlainObject(sourceNode.ast));
        // After generating the stylesheet, parse again (so line numbers are correct)
        sourceNode.ast = window.csstree.toPlainObject((window.csstree.parse(sourceNode.stylesheet.content, { positions: true, parseValue: false })));
        this.stylesheetChanged.emit({ name: sourceNode.stylesheet.name, newStyle: sourceNode.stylesheet.content, oldStyle: old, changeSource: 'styleupdate' });
        return true;
    }
    removeDeclarationFromRule(rule, property) {
        let index = rule.ast.block.children.findIndex(x => x.property == property);
        if (index == -1)
            return false;
        rule.ast.block.children.splice(index, 1);
        const ss = this._stylesheets.get(rule.stylesheetName);
        const old = ss.stylesheet.content;
        ss.stylesheet.content = window.csstree.generate(window.csstree.fromPlainObject(this._stylesheets.get(rule.stylesheetName).ast));
        // After generating the stylesheet, parse again (so line numbers are correct)
        ss.ast = window.csstree.toPlainObject((window.csstree.parse(this._stylesheets.get(rule.stylesheetName).stylesheet.content, { positions: true, parseValue: false })));
        this.stylesheetChanged.emit({ name: ss.stylesheet.name, newStyle: ss.stylesheet.content, oldStyle: old, changeSource: 'styleupdate' });
        return true;
    }
    updateCompleteStylesheet(name, newStyle) {
        const ss = this._stylesheets.get(name);
        if (ss.stylesheet.content != newStyle) {
            const old = ss.stylesheet.content;
            ss.stylesheet.content = newStyle;
            this.stylesheetChanged.emit({ name: ss.stylesheet.name, newStyle: ss.stylesheet.content, oldStyle: old, changeSource: 'styleupdate' });
        }
    }
    /* Section covers the internal traversal creation of rules and declarations */
    *rulesFromAST(ast, stylesheet, source, designItem, previousCheck = '') {
        let parent = ast["children"] != null ? ast : ast.block;
        for (const child of parent.children) {
            if (child.type == "Atrule") {
                const currentCheck = this.buildAtRuleString(child, stylesheet);
                if (currentCheck.type === "@media" && !window.matchMedia(currentCheck.sel))
                    continue;
                if (currentCheck.type === "@supports" && !CSS.supports(currentCheck.sel))
                    continue;
                let ruleCollection = this.rulesFromAST(child, stylesheet, source, designItem, previousCheck + currentCheck.type + " " + currentCheck.sel + "\n");
                if (ruleCollection) {
                    for (const r of ruleCollection) {
                        if (!this.elementMatchesASelector(designItem, this.buildSelectorString(r.ast.prelude, stylesheet)))
                            continue;
                        yield r;
                    }
                }
            }
            if (child.type == "Rule") {
                let selectors = this.buildSelectorString(child.prelude, stylesheet);
                if (!this.elementMatchesASelector(designItem, selectors))
                    continue;
                yield ({
                    ast: child,
                    selector: previousCheck + this.buildSelectorString(child.prelude, stylesheet).join(", "),
                    specificity: this.getSpecificity(child.prelude),
                    stylesheetName: source,
                    declarations: null,
                });
            }
        }
        ;
    }
    /* Utility functions for building selectors, specificity and so on  */
    astHasChildren(ast) {
        return ast != null && ast["children"] != null && ast["children"].length > 0;
    }
    buildSelectorString(selectorsAST, stylesheet) {
        if (selectorsAST.type == 'SelectorList') {
            return [...selectorsAST.children.map(x => this.buildSelectorString(x, stylesheet)[0])];
        }
        else {
            return [stylesheet.substring(selectorsAST.loc.start.offset, selectorsAST.loc.end.offset)];
        }
    }
    getSpecificity(selector) {
        const specificities = calculateSpecifity(selector);
        let sum = 0;
        specificities.forEach(specificity => sum += specificity.a * 10000 + specificity.b * 100 + specificity.c);
        return sum;
    }
    findDeclarationInRule(rule, styleName) {
        return rule.block.children.find(declaration => declaration.property == styleName);
    }
    buildAtRuleString(ast, stylesheet) {
        return {
            sel: stylesheet.slice(ast.prelude.loc.start.offset, ast.prelude.loc.end.offset),
            type: "@" + ast.name
        };
    }
    sortDeclarations(declarations) {
        if (declarations == null || declarations.length == 0)
            return null;
        return declarations.sort((dec1, dec2) => {
            if (dec1.parent.specificity > dec2.parent.specificity)
                return -1;
            return 1;
        });
    }
}
