// @node-projects/web-component-designer/dist/elements/services/undoService/transactionItems/AttributeChangeAction.js
export class AttributeChangeAction {
  constructor(designItem, name, newValue, oldValue) {
    this.title = "Change Attribute";
    this.designItem = designItem;
    this.name = name;
    this.newValue = newValue;
    this.oldValue = oldValue;
  }

  get affectedItems() {
    return [this.designItem];
  }

  undo() {
    if (this.oldValue == null) {
      this.designItem.attributes.delete(this.name);
      this.designItem.element.removeAttribute(this.name);
    } else {
      this.designItem.attributes.set(this.name, this.oldValue);

      if (this.name != "draggable") {
        this.designItem.element.setAttribute(this.name, this.oldValue);
      }
    }
  }

  do() {
    if (this.newValue == null) {
      this.designItem.attributes.delete(this.name);
      this.designItem.element.removeAttribute(this.name);
    } else {
      this.designItem.attributes.set(this.name, this.newValue);

      if (this.name != "draggable") {
        this.designItem.element.setAttribute(this.name, this.newValue);
      }
    }
  }

  mergeWith(other) {
    if (other instanceof AttributeChangeAction && this.designItem === other.designItem && this.name === other.name) {
      this.newValue = other.newValue;
      return true;
    }

    return false;
  }

}
