// @node-projects/web-component-designer/dist/elements/services/undoService/transactionItems/DeleteAction.js
import { DomHelper } from '/web-component-designer-demo/node_modules/@node-projects/base-custom-webcomponent/dist/DomHelper.js';
export class DeleteAction {
    constructor(deletedItems, extensionManager) {
        this.title = "Delete Items";
        this.deletedItems = deletedItems;
        this.extensionManager = extensionManager;
    }
    title;
    get affectedItems() {
        return this.deletedItems;
    }
    undo() {
        for (let n = 0; n < this.deletedItems.length; n++) {
            this._parentItems[n].insertChild(this.deletedItems[n], this._parentIndexes[n]);
        }
    }
    do() {
        this._parentItems = [];
        this._parentIndexes = [];
        for (let n = 0; n < this.deletedItems.length; n++) {
            this._parentItems.push(this.deletedItems[n].parent);
            this._parentIndexes.push(DomHelper.nodeIndex(this.deletedItems[n].element));
        }
        for (let n = 0; n < this.deletedItems.length; n++) {
            this.deletedItems[n].remove();
        }
        this.extensionManager.removeExtensions(this.deletedItems);
    }
    deletedItems;
    extensionManager;
    _parentItems;
    _parentIndexes;
    mergeWith(other) {
        return false;
    }
}
