// @node-projects/web-component-designer/dist/elements/services/undoService/transactionItems/DeleteAction.js
import { DomHelper } from "../../../../../../base-custom-webcomponent/dist/DomHelper.js";
export class DeleteAction {
  constructor(deletedItems) {
    this.title = "Delete Items";
    this.deletedItems = deletedItems;
  }

  get affectedItems() {
    return this.deletedItems;
  }

  undo() {
    for (let n = 0; n < this.deletedItems.length; n++) {
      if (this._parentIndexes[n] == 0) this._parentItems[n].element.insertAdjacentElement('afterbegin', this.deletedItems[n].element);else {
        let el = this._parentItems[n].element.children[this._parentIndexes[n] - 1];
        el.insertAdjacentElement('afterend', this.deletedItems[n].element);
      }
    }
  }

  do() {
    this._parentItems = [];
    this._parentIndexes = [];

    for (let n = 0; n < this.deletedItems.length; n++) {
      this._parentItems.push(this.deletedItems[n].parent);

      this._parentIndexes.push(DomHelper.nodeIndex(this.deletedItems[n].element));
    }

    for (let n = 0; n < this.deletedItems.length; n++) {
      this.deletedItems[n].element.remove();
    }
  }

  mergeWith(other) {
    return false;
  }

}
