// @node-projects/web-component-designer/dist/elements/widgets/designerView/extensions/AbstractExtension.js
export class AbstractExtension {
    constructor(extensionManager, designerView, extendedItem) {
        this.overlays = [];
        this.extensionManager = extensionManager;
        this.designerView = designerView;
        this.extendedItem = extendedItem;
        this.overlayLayerView = designerView.overlayLayer;
    }
    _removeAllOverlays() {
        for (let o of this.overlays) {
            try {
                this.overlayLayerView.removeOverlay(o);
            }
            catch (err) {
                console.error(err);
            }
        }
        this.overlays = [];
    }
    _drawLineOverlay(x1, y1, x2, y2, className, line, overlayLayer) {
        if (!line) {
            line = document.createElementNS("http://www.w3.org/2000/svg", "line");
            this.overlayLayerView.addOverlay(line, overlayLayer);
            this.overlays.push(line);
        }
        line.setAttribute('x1', x1);
        line.setAttribute('y1', y1);
        line.setAttribute('x2', x2);
        line.setAttribute('y2', y2);
        if (className)
            line.setAttribute('class', className);
        return line;
    }
    _drawCircleOverlay(x, y, radius, className, circle, overlayLayer) {
        if (!circle) {
            circle = document.createElementNS("http://www.w3.org/2000/svg", "circle");
            this.overlayLayerView.addOverlay(circle, overlayLayer);
            this.overlays.push(circle);
        }
        circle.setAttribute('cx', x);
        circle.setAttribute('cy', y);
        circle.setAttribute('r', radius);
        if (className)
            circle.setAttribute('class', className);
        return circle;
    }
    _drawRect(x, y, w, h, className, rect, overlayLayer) {
        if (!rect) {
            rect = document.createElementNS("http://www.w3.org/2000/svg", "rect");
            this.overlayLayerView.addOverlay(rect, overlayLayer);
            this.overlays.push(rect);
        }
        rect.setAttribute('x', x);
        rect.setAttribute('y', y);
        rect.setAttribute('width', w);
        rect.setAttribute('height', h);
        if (className)
            rect.setAttribute('class', className);
        return rect;
    }
    _drawText(text, x, y, className, textEl, overlayLayer) {
        if (!textEl) {
            textEl = document.createElementNS("http://www.w3.org/2000/svg", "text");
            this.overlayLayerView.addOverlay(textEl, overlayLayer);
            this.overlays.push(textEl);
        }
        textEl.setAttribute('x', x);
        textEl.setAttribute('y', y);
        textEl.textContent = text;
        if (className)
            textEl.setAttribute('class', className);
        return textEl;
    }
}
