// @node-projects/web-component-designer/dist/elements/services/propertiesService/propertyEditors/BasePropertyEditor.js
import { ValueType } from '../ValueType.js';
import { PropertyChangeAction } from '../../undoService/transactionItems/PropertyChangeAction.js';
export class BasePropertyEditor {
    constructor(property) {
        this.disableChangeNotification = false;
        this.property = property;
    }
    _valueChanged(newValue) {
        if (!this.disableChangeNotification) {
            if (this.designItems && this.designItems.length) {
                const cg = this.designItems[0].openGroup("set property: " + this.property.name);
                for (let d of this.designItems) {
                    const oldValue = this.property.service.getValue([d], this.property);
                    const action = new PropertyChangeAction(d, this.property, newValue, oldValue);
                    d.instanceServiceContainer.undoService.execute(action);
                }
                cg.commit();
            }
        }
    }
    designItemsChanged(designItems) {
        this.designItems = designItems;
    }
    refreshValueWithoutNotification(valueType, value) {
        if (valueType == ValueType.none)
            this.element.classList.add('unset-value');
        else
            this.element.classList.remove('unset-value');
        this.disableChangeNotification = true;
        this.refreshValue(valueType, value);
        this.disableChangeNotification = false;
    }
}
