// @node-projects/web-component-designer/dist/elements/services/ServiceContainer.js
import { BaseServiceContainer } from "./BaseServiceContainer.js";
import { CodeViewMonaco } from "../widgets/codeView/code-view-monaco.js";
import { DemoView } from "../widgets/demoView/demoView.js";
import { GlobalContext } from "./GlobalContext.js";
export class ServiceContainer extends BaseServiceContainer {
  constructor() {
    super(...arguments);
    this.config = {
      codeViewWidget: CodeViewMonaco,
      demoViewWidget: DemoView
    };
    this.designerExtensions = new Map();
    this.globalContext = new GlobalContext();
    this.designerTools = new Map();
  }

  get bindingServices() {
    return this.getServices('bindingService');
  }

  get propertiesServices() {
    return this.getServices('propertyService');
  }

  get containerServices() {
    return this.getServices('containerService');
  }

  get snaplinesProviderService() {
    return this.getLastService('snaplinesProviderService');
  }

  get elementsServices() {
    return this.getServices('elementsService');
  }

  get instanceServices() {
    return this.getServices('instanceService');
  }

  get editorTypesServices() {
    return this.getServices('editorTypesService');
  }

  get htmlWriterServices() {
    return this.getServices('htmlWriterService');
  }

  get htmlParserService() {
    return this.getLastService('htmlParserService');
  }

  get intializationService() {
    return this.getLastService('intializationService');
  }

  get elementAtPointService() {
    return this.getLastService('elementAtPointService');
  }

}
